zanata-war/src/main/java/org/zanata/service/impl/MergeTranslationsServiceImpl.java

```diff
@@ -0,0 +1,278 @@
+            return AsyncTaskResult.taskResult();
+        }
+
+        if (getSupportedLocales(targetProjectSlug, targetVersionSlug).isEmpty()) {
+            log.error("No locales enabled in target version of {} [{}]",
+                    targetProjectSlug, targetVersionSlug);
+            return AsyncTaskResult.taskResult();
+        }
+
+        Optional<MergeTranslationsTaskHandle> taskHandleOpt =
+                Optional.fromNullable(handle);
+
+        if (taskHandleOpt.isPresent()) {
+            prepareMergeTranslationsHandle(sourceVersion, targetVersion,
+                    taskHandleOpt.get());
+        }
+
+        Stopwatch overallStopwatch = Stopwatch.createStarted();
+        log.info("merge translations start: from {} to {}", sourceProjectSlug
+                + ":" + sourceVersionSlug, targetProjectSlug + ":"
+                + targetVersionSlug);
+
+        int startCount = 0;
+        int totalCount = getTotalMatchCount(sourceVersion.getId(),
+                targetVersion.getId());
+
+        List<HLocale> supportedLocales = getSupportedLocales(targetVersion
+                .getProject().getSlug(), targetVersion.getSlug());
+
+        while (startCount < totalCount) {
+            int processedCount =
+                    mergeTranslationBatch(sourceVersion, targetVersion,
+                            supportedLocales, useNewerTranslation, startCount,
+                            TRANSLATION_BATCH_SIZE);
+            if (taskHandleOpt.isPresent()) {
+                taskHandleOpt.get().increaseProgress(processedCount);
+            }
+
+            startCount += TRANSLATION_BATCH_SIZE;
+            textFlowDAO.clear();
+        }
+        versionStateCacheImpl.clearVersionStatsCache(targetVersion.getId());
+        log.info("merge translation end: from {} to {}, {}", sourceProjectSlug
+                + ":" + sourceVersionSlug, targetProjectSlug + ":"
+                + targetVersionSlug, overallStopwatch);
+
+        return AsyncTaskResult.taskResult();
+    }
+
+    protected int mergeTranslationBatch(HProjectIteration sourceVersion,
+            HProjectIteration targetVersion, List<HLocale> supportedLocales,
+            boolean useNewerTranslation, int offset, int batchSize) {
+        try {
+            return new MergeTranslationsWork(sourceVersion.getId(),
+                    targetVersion.getId(), offset, batchSize,
+                    useNewerTranslation, supportedLocales,
+                    textFlowDAO, translationStateCacheImpl)
+                    .workInTransaction();
+        } catch (Exception e) {
+            log.warn("exception during copy text flow target", e);
+            return 0;
+        }
+    }
+
+    /**
+     * Check if sourceVersion or targetVersion has source document.
+     *
+     * @param sourceVersion
+     * @param targetVersion
+     */
+    private boolean isVersionsEmpty(HProjectIteration sourceVersion,
+            HProjectIteration targetVersion) {
```

Comment:
> If you really want to copy another one they can add you as a maintainer.

I disagree with this for the same reason I would not want to give someone commit access on my github projects just so they can fork them.